Transition from Coinbase Pro to Coinbase Advanced broker (API v3)

marcelonyc opened the pull request:
# Description
In June Coinbase announced that they will begin sunsetting Coinbase Pro by end of 2022. 

https://www.coinbase.com/blog/hello-advanced-trade-goodbye-coinbase-pro

This PR implements integration with their new Coinbase Advanced API. This API (v3) is still under development and it is significantly different than the Pro API. It also requires using different APIs depending on the task. 

- [x] Screenshot of the feature or the bug before/after fix, if applicable.
![orders](https://user-images.githubusercontent.com/12512887/206734024-75709a87-9275-4f61-a2ae-7981b6b5dd04.PNG)
![accounts](https://user-images.githubusercontent.com/12512887/206734062-a10a5daf-ae59-4c83-91f0-0c52a01ee964.PNG)

- [x] Relevant motivation and context.
Coinbase sunsetting PRO API
- [x] List any dependencies that are required for this change.
- Added environment variables for API KEYS keys_model.py
- Added menu text for new paths en.yml
- Added a function to validate datetime arguments to helper_funcs.py


# How has this been tested?
I manually tested each command including placing/canceling orders with Coinbase Advance. Order commands have a --dry-run option to review the json payload. Screenshot of each order test here:
![dry-run](https://user-images.githubusercontent.com/12512887/208433931-91f2ba9b-a00d-4e8d-810a-b991c22a3373.png)
![cancelorder](https://user-images.githubusercontent.com/12512887/208433478-6a9c3fce-0e55-4bcf-b275-9acc77c76952.png)
![limitgtcbuy](https://user-images.githubusercontent.com/12512887/208433487-5692d602-1b90-479b-b742-b80922f78372.png)
![limitgtcsell](https://user-images.githubusercontent.com/12512887/208433494-db924aaf-29c6-48fb-8167-180801e2f3ed.png)
![limitgtdbuy](https://user-images.githubusercontent.com/12512887/208433517-19012626-1955-47c3-b9bb-1a6d98d2e8b8.png)
![limitgtd_sell](https://user-images.githubusercontent.com/12512887/208433527-c90f6a79-ab15-40ea-ac18-00b15d947afa.png)
![stopgtcbuy](https://user-images.githubusercontent.com/12512887/208433547-ad17a858-5394-4c0e-b543-22f5b2deb786.png)
![stopgtcsell](https://user-images.githubusercontent.com/12512887/208433566-233cf812-cb07-4032-b67f-2580c6d22da2.png)
![stopgtdbuy](https://user-images.githubusercontent.com/12512887/208433576-e21b5224-931b-4f92-8e2b-ee7f1d7466f2.png)
![stopgtdsell](https://user-images.githubusercontent.com/12512887/208433595-cd2f45ba-c9be-4c2d-899b-69bef8e0347f.png)

* Please describe the tests that you ran to verify your changes.
* Provide instructions so we can reproduce.
* Please also list any relevant details for your test configuration.
- [x] Make sure affected commands still run in terminal
- [ ] Ensure the SDK still works
- [ ] Check any related reports


# Checklist:

- [ ] Update [our Hugo documentation](https://openbb-finance.github.io/OpenBBTerminal/) following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/website).
- Not done yet
- [ ] Update our tests following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/tests).
- Partial
- [x] Make sure you are following our [CONTRIBUTING guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/blob/main/CONTRIBUTING.md).
- [ ] If a feature was added make sure to add it to the corresponding [scripts file](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/scripts).


# Others
- [X] I have performed a self-review of my own code.
- [X] I have commented my code, particularly in hard-to-understand areas.

jmaslek commented:
Followup:  havent been ignoring this.  I just switched my account over to this new v3 thing.  Said I would have access in 2 days lol.  So will be able to test out then

marcelonyc commented:
@jmaslek no rush. I am taking the time to clean up the code and also adding create/cancel order menus. Unfortunately, Coinbase is not providing a sandbox environment for the Advanced Trading API (Yet). I have to test my orders code carefully so I do not wipe out my portfolio or get an unexpected tax bill :)

jmaslek commented:
Hey I am finally giving this a go.  I got my api key up and whatnot:
```
│     coinbase_advanced      CB_ADV                     defined, test passed     
```

But when I try to enter the cbadv menu:
```
023 Jan 05, 11:28 (🦋) /portfolio/bro/ $ cbadv

2023 Jan 05, 11:28 (🦋) /portfolio/bro/cbadv/ $ ?
```

Nothing is happening.  So I look at the debug printout and I am getting:
```
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/portfolio/brokers/bro_controller.py", line 80, in call_cbadv
    self.queue = self.load_class(
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/parent_classes.py", line 219, in load_class
    return class_ins(*args, **kwargs).menu()
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/portfolio/brokers/coinbase_advanced/coinbase_advanced_controller.py", line 70, in __init__
    choices: dict = self.choices_default
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/parent_classes.py", line 115, in choices_default
    choices = build_controller_choice_map(controller=self)
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/core/completer/choices.py", line 322, in build_controller_choice_map
    raise Exception(
Exception: On command : `cancelorder`.
'status'
```

This has to do with autocomplete -- but everything on your end looks good.  IS this happening for you at all?

jmaslek commented on openbb_terminal/cryptocurrency/coinbase_advanced_helpers.py:
```diff
@@ -0,0 +1,423 @@
+        flat_response.append(_flat_response)
+
+    df = pd.DataFrame(flat_response)
+    if df.empty:
```
I think this is causing my problem.  You may need to define the columns like you do above?
```
columns=[
                "product_id",
                "side",
                "price",
                "size",
                "order_type",
                "created_time",
                "status",
            ]
```

jmaslek replied:
These empty dataframes also need an order_id column

marcelonyc replied:
@jmaslek What is the command sequence that raises the error?

jmaslek replied:
This happens when I just try entering the menu.  its from the argument choices (which we use to autogenerate the autocomplete)

marcelonyc replied:
I see now. The choices for the argument order_id  get loaded from a CB query. The choices get generated when you first enter the menu. I retrieve all the orders first and then filter. I think that is why it works for me, I have some orders. I am rethinking this, I will make it a string and you have to paste the order_id.

jmaslek commented:
This is probably a me issue as well in the sense of maybe I didnt do coinbase advanced setup correctly, but the account is showing no response even though i have a small amount of some coins like btc/eth
```

2023 Jan 05, 11:44 (🦋) /portfolio/bro/cbadv/ $ account


2023 Jan 05, 11:44 (🦋) /portfolio/bro/cbadv/ $ account
```

jmaslek commented:
> @jmaslek I am not sure why it would come back empty if you have valid assets. Unless the assets you have are "Not Tradable in your region". For some reason, the request does not return those For example in the UI (for my account) XRP shows as "Not Tradeble in your region". I have a few XRP in Coinbase, they do not show when I run accounts.

Hmm.  Interesting.  Im in the US, but I have BTC/ETH/ETH2 so I assume those would show up.

marcelonyc commented:
> 
@jmaslek  when you created the API key did you select all the accounts or just My Wallet? 
![Screenshot from 2023-01-05 13-09-02](https://user-images.githubusercontent.com/12512887/210850506-5c338b1a-d56f-40ac-a9ba-7d82563535cc.png)

jmaslek commented:
> > 
> 
> @jmaslek when you created the API key did you select all the accounts or just My Wallet? ![Screenshot from 2023-01-05 13-09-02](https://user-images.githubusercontent.com/12512887/210850506-5c338b1a-d56f-40ac-a9ba-7d82563535cc.png)

Do not remember.  Knowing meself I did all

marcelonyc commented:
I changed the cancelorder command to avoid the error @jmaslek saw. And btw, I realized that at some point, I removed ```cancelorder``` from the menu. It is there now. 

For ```cancelorder```, Instead of loading a list of orders, which does not work well,  the user has to paste the ```order_id``` ( ```cancelorder --order_id <paste order id here> ```)

To make the order_id visible when users list orders, I gave the ```orders``` command a ```--slim``` argument to display less columns.

![Screen Shot 2023-01-05 at 4 28 14 PM](https://user-images.githubusercontent.com/12512887/210883374-004f36c9-4484-44a1-8beb-11658d9e3720.png)

A side note on this. If print_rich_table had the option to provide columns min/max width, I could just changed the width of the order_id column to be wider.  Maybe I will make this a feature request. 

I am not sure why @jmaslek can't list accounts. I was able to reproduce it with an API key with limited account access, but I am not sure if that is the case he ran into.
